python27Packages.scandir: test failures on aarch64-darwin

I noticed ofborg was [failing](https://logs.nix.ci/?key=nixos/nixpkgs.155251&attempt_id=c5bd7269-fc6f-42a4-8a57-636461c14f3a) to build resholve (which has no short-run path off of python27) on aarch64-darwin due to test failures in scandir. 

I don't expect there to be much enthusiasm for the exact failure here, but I'm reporting it in case the root cause is meaningful...

### Describe the bug
<details>
<summary>On aarch64-darwin, 1 scandir test errors and 5 fail as follows:</summary>

```console
Sourcing python-remove-tests-dir-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing setuptools-build-hook
Using setuptoolsBuildPhase
Using setuptoolsShellHook
Sourcing pip-install-hook
Using pipInstallPhase
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
Sourcing setuptools-check-hook
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/s0l2vkcm04z0r64ia0xh6cn91ly8xiwr-scandir-1.10.0.tar.gz
source root is scandir-1.10.0
setting SOURCE_DATE_EPOCH to timestamp 1552154153 of file scandir-1.10.0/setup.cfg
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "configurePhase" }
configuring
no configure script, doing nothing
@nix { "action": "setPhase", "phase": "buildPhase" }
building
Executing setuptoolsBuildPhase
/nix/store/7nv2sjvzxccfvrca2a9i0axxhm9y3zxp-python-2.7.18/lib/python2.7/distutils/extension.py:133: UserWarning: Unknown Extension options: 'optional'
  warnings.warn(msg)
running bdist_wheel
running build
running build_py
creating build
creating build/lib.macosx-11.0-arm64-2.7
copying scandir.py -> build/lib.macosx-11.0-arm64-2.7
running build_ext
building '_scandir' extension
creating build/temp.macosx-11.0-arm64-2.7
clang -I/nix/store/7nv2sjvzxccfvrca2a9i0axxhm9y3zxp-python-2.7.18/include/python2.7 -c _scandir.c -o build/temp.macosx-11.0-arm64-2.7/_scandir.o
clang -bundle -undefined dynamic_lookup build/temp.macosx-11.0-arm64-2.7/_scandir.o -L/nix/store/7nv2sjvzxccfvrca2a9i0axxhm9y3zxp-python-2.7.18/lib -o build/lib.macosx-11.0-arm64-2.7/_scandir.so
installing to build/bdist.macosx-11.0-arm64/wheel
running install
running install_lib
creating build/bdist.macosx-11.0-arm64
creating build/bdist.macosx-11.0-arm64/wheel
copying build/lib.macosx-11.0-arm64-2.7/_scandir.so -> build/bdist.macosx-11.0-arm64/wheel
copying build/lib.macosx-11.0-arm64-2.7/scandir.py -> build/bdist.macosx-11.0-arm64/wheel
running install_egg_info
running egg_info
writing scandir.egg-info/PKG-INFO
writing top-level names to scandir.egg-info/top_level.txt
writing dependency_links to scandir.egg-info/dependency_links.txt
reading manifest file 'scandir.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'scandir.egg-info/SOURCES.txt'
Copying scandir.egg-info to build/bdist.macosx-11.0-arm64/wheel/scandir-1.10.0-py2.7.egg-info
running install_scripts
adding license file "LICENSE.txt" (matched pattern "LICEN[CS]E*")
creating build/bdist.macosx-11.0-arm64/wheel/scandir-1.10.0.dist-info/WHEEL
creating 'dist/scandir-1.10.0-cp27-cp27mu-macosx_11_0_arm64.whl' and adding 'build/bdist.macosx-11.0-arm64/wheel' to it
adding '_scandir.so'
adding 'scandir.py'
adding 'scandir-1.10.0.dist-info/LICENSE.txt'
adding 'scandir-1.10.0.dist-info/METADATA'
adding 'scandir-1.10.0.dist-info/WHEEL'
adding 'scandir-1.10.0.dist-info/top_level.txt'
adding 'scandir-1.10.0.dist-info/RECORD'
removing build/bdist.macosx-11.0-arm64/wheel
Finished executing setuptoolsBuildPhase
@nix { "action": "setPhase", "phase": "installPhase" }
installing
Executing pipInstallPhase
/private/tmp/nix-build-python2.7-scandir-1.10.0.drv-0/scandir-1.10.0/dist /private/tmp/nix-build-python2.7-scandir-1.10.0.drv-0/scandir-1.10.0
[33mDEPRECATION: Python 2.7 reached the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 is no longer maintained. pip 21.0 will drop support for Python 2.7 in January 2021. More details about Python 2 support in pip can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support pip 21.0 will remove support for this functionality.[0m
Processing ./scandir-1.10.0-cp27-cp27mu-macosx_11_0_arm64.whl
Installing collected packages: scandir
Successfully installed scandir-1.10.0
/private/tmp/nix-build-python2.7-scandir-1.10.0.drv-0/scandir-1.10.0
Finished executing pipInstallPhase
@nix { "action": "setPhase", "phase": "fixupPhase" }
post-installation fixup
strip is /nix/store/ljxsc2chsz0qr6318mpnwhc3cagcfck7-clang-wrapper-11.1.0/bin/strip
stripping (with command strip and flags -S) in /nix/store/p5dfng0h9gaji94fvhlpbr2bnfv92kgq-python2.7-scandir-1.10.0/lib 
patching script interpreter paths in /nix/store/p5dfng0h9gaji94fvhlpbr2bnfv92kgq-python2.7-scandir-1.10.0
Executing pythonRemoveTestsDir
Finished executing pythonRemoveTestsDir
@nix { "action": "setPhase", "phase": "installCheckPhase" }
running install tests
test_basic (test_scandir.TestScandirC) ... ok
test_bytes (test_scandir.TestScandirC) ... ok
test_dir_entry (test_scandir.TestScandirC) ... ok
test_file_attributes (test_scandir.TestScandirC) ... skipped 'st_file_attributes not supported'
test_path (test_scandir.TestScandirC) ... ok
test_returns_iter (test_scandir.TestScandirC) ... ok
test_stat (test_scandir.TestScandirC) ... ok
test_symlink (test_scandir.TestScandirC) ... ok
test_unicode (test_scandir.TestScandirC) ... ok
test_walk_unicode_handling (test_scandir.TestScandirC) ... ok
test_iter_returns_dir_entry (test_scandir.TestScandirDirEntry) ... ok
test_basic (test_scandir.TestScandirGeneric) ... ok
test_bytes (test_scandir.TestScandirGeneric) ... ok
test_dir_entry (test_scandir.TestScandirGeneric) ... ok
test_file_attributes (test_scandir.TestScandirGeneric) ... skipped 'st_file_attributes not supported'
test_path (test_scandir.TestScandirGeneric) ... ok
test_returns_iter (test_scandir.TestScandirGeneric) ... ok
test_stat (test_scandir.TestScandirGeneric) ... ok
test_symlink (test_scandir.TestScandirGeneric) ... ok
test_unicode (test_scandir.TestScandirGeneric) ... ok
test_walk_unicode_handling (test_scandir.TestScandirGeneric) ... ok
test_basic (test_scandir.TestScandirPython) ... FAIL
test_bytes (test_scandir.TestScandirPython) ... FAIL
test_dir_entry (test_scandir.TestScandirPython) ... ERROR
test_file_attributes (test_scandir.TestScandirPython) ... skipped 'st_file_attributes not supported'
test_path (test_scandir.TestScandirPython) ... FAIL
test_returns_iter (test_scandir.TestScandirPython) ... ok
test_stat (test_scandir.TestScandirPython) ... ok
test_symlink (test_scandir.TestScandirPython) ... FAIL
test_unicode (test_scandir.TestScandirPython) ... FAIL
test_walk_unicode_handling (test_scandir.TestScandirPython) ... ok
test_traversal (test_walk.TestWalk) ... ok
test_symlink_to_directory (test_walk.TestWalkSymlink) ... ok
test_symlink_to_file (test_walk.TestWalkSymlink) ... ok

======================================================================
ERROR: test_dir_entry (test_scandir.TestScandirPython)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/tmp/nix-build-python2.7-scandir-1.10.0.drv-0/scandir-1.10.0/test/test_scandir.py", line 113, in test_dir_entry
    e = entries['file1.txt']
KeyError: u'file1.txt'

======================================================================
FAIL: test_basic (test_scandir.TestScandirPython)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/tmp/nix-build-python2.7-scandir-1.10.0.drv-0/scandir-1.10.0/test/test_scandir.py", line 107, in test_basic
    ('linkdir', True), ('subdir', True)])
AssertionError: Lists differ: [(u'', True), (u'', True), (u'... != [(u'file1.txt', False), (u'fil...

First differing element 0:
(u'', True)
(u'file1.txt', False)

First list contains 2 additional elements.
First extra element 4:
(u'', True)

- [(u'', True), (u'', True), (u'', True), (u'', True), (u'', True), (u'', True)]
+ [(u'file1.txt', False),
+  (u'file2.txt', False),
+  (u'linkdir', True),
+  (u'subdir', True)]

======================================================================
FAIL: test_bytes (test_scandir.TestScandirPython)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/tmp/nix-build-python2.7-scandir-1.10.0.drv-0/scandir-1.10.0/test/test_scandir.py", line 211, in test_bytes
    self.assertEqual(len(entries), 1)
AssertionError: 0 != 1

======================================================================
FAIL: test_path (test_scandir.TestScandirPython)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/tmp/nix-build-python2.7-scandir-1.10.0.drv-0/scandir-1.10.0/test/test_scandir.py", line 168, in test_path
    ['file1.txt', 'file2.txt', 'linkdir', 'subdir'])
AssertionError: Lists differ: [u'', u'', u'', u'', u'', u''] != [u'file1.txt', u'file2.txt', u...

First differing element 0:
u''
u'file1.txt'

First list contains 2 additional elements.
First extra element 4:
u''

- [u'', u'', u'', u'', u'', u'']
+ [u'file1.txt', u'file2.txt', u'linkdir', u'subdir']

======================================================================
FAIL: test_symlink (test_scandir.TestScandirPython)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/tmp/nix-build-python2.7-scandir-1.10.0.drv-0/scandir-1.10.0/test/test_scandir.py", line 183, in test_symlink
    ('linksubdir', False)])
AssertionError: Lists differ: [(u'', False), (u'', False), (... != [(u'file1.txt', False), (u'lin...

First differing element 0:
(u'', False)
(u'file1.txt', False)

First list contains 2 additional elements.
First extra element 4:
(u'', False)

- [(u'', False),
+ [(u'file1.txt', False),
?     +++++++++

+  (u'link_to_dir', True),
+  (u'link_to_file', True),
+  (u'linksubdir', False)]
-  (u'', False),
-  (u'', False),
-  (u'', False),
-  (u'', False),
-  (u'', False)]

======================================================================
FAIL: test_unicode (test_scandir.TestScandirPython)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/tmp/nix-build-python2.7-scandir-1.10.0.drv-0/scandir-1.10.0/test/test_scandir.py", line 229, in test_unicode
    self.assertEqual(len(entries), 1)
AssertionError: 0 != 1

----------------------------------------------------------------------
Ran 34 tests in 0.012s

FAILED (failures=5, errors=1, skipped=3)
```
</details>

### Additional context
I don't see any similar issues upstream. I haven't yet filed anything there (but I do have a tab open to do so, if it sounds like this is intrinsic).

### Notify maintainers

@abbradar Just in case you've seen anything like this?

--- a/pkgs/development/misc/resholve/resholve.nix
+++ b/pkgs/development/misc/resholve/resholve.nix
@@ -19,24 +19,16 @@ python27Packages.buildPythonApplication {
 
   propagatedBuildInputs = [
     oildev
-    /*
-    Disable configargparse's tests on aarch64-darwin.
-    Several of py27 scandir's tests fail on aarch64-darwin. Chain:
-    configargparse -> pytest-check-hook -> pytest -> pathlib2 -> scandir
-    TODO: drop if https://github.com/NixOS/nixpkgs/issues/156807 resolves?
-    */
-    (python27Packages.configargparse.overridePythonAttrs (old: {
-      doCheck = stdenv.hostPlatform.system != "aarch64-darwin";
-    }))
+    python27Packages.configargparse
   ];
 
-  patchPhase = ''
+  postPatch = ''
     for file in setup.cfg _resholve/version.py; do
       substituteInPlace $file --subst-var-by version ${version}
     done
   '';
 
-   postInstall = ''
+  postInstall = ''
     installManPage resholve.1
   '';
 

--- a/pkgs/development/python-modules/pygobject/default.nix
+++ b/pkgs/development/python-modules/pygobject/default.nix
@@ -1,4 +1,4 @@
-{ lib, stdenv, fetchurl, python, buildPythonPackage, pkg-config, glib, isPy3k, pythonAtLeast }:
+{ lib, stdenv, fetchurl, fetchpatch, python, buildPythonPackage, pkg-config, glib, isPy3k, pythonAtLeast }:
 
 buildPythonPackage rec {
   pname = "pygobject";
@@ -15,6 +15,11 @@ buildPythonPackage rec {
 
   patches = lib.optionals stdenv.isDarwin [
     ./pygobject-2.0-fix-darwin.patch
+    (fetchpatch {
+      url = "https://github.com/macports/macports-ports/raw/f2975d5bbbc2459c661905c5a850cc661fa32f55/python/py-gobject/files/py-gobject-dynamic_lookup-11.patch";
+      sha256 = "sha256-mtlyu+La3+iC5iQAmVJzDA5E35XGaRQy/EKXzvrWRCg=";
+      extraPrefix = "";
+    })
   ];
 
   configureFlags = [ "--disable-introspection" ];

--- a/pkgs/development/python2-modules/pygtk/default.nix
+++ b/pkgs/development/python2-modules/pygtk/default.nix
@@ -36,7 +36,8 @@ buildPythonPackage rec {
 
   buildPhase = "buildPhase";
 
-  NIX_CFLAGS_COMPILE = lib.optionalString stdenv.isDarwin "-ObjC";
+  NIX_CFLAGS_COMPILE = lib.optional stdenv.isDarwin "-ObjC"
+    ++ lib.optional (stdenv.isDarwin && stdenv.isAarch64) "-lpython2.7";
 
   installPhase = "installPhase";
 

--- a/pkgs/development/python2-modules/scandir/add-aarch64-darwin-dirent.patch
+++ b/pkgs/development/python2-modules/scandir/add-aarch64-darwin-dirent.patch
@@ -1,0 +1,28 @@
+diff --git a/scandir.py b/scandir.py
+index 3f602fb..40af3e5 100644
+--- a/scandir.py
++++ b/scandir.py
+@@ -23,6 +23,7 @@ from os import listdir, lstat, stat, strerror
+ from os.path import join, islink
+ from stat import S_IFDIR, S_IFLNK, S_IFREG
+ import collections
++import platform
+ import sys
+ 
+ try:
+@@ -432,6 +433,15 @@ elif sys.platform.startswith(('linux', 'darwin', 'sunos5')) or 'bsd' in sys.plat
+                     ('__d_padding', ctypes.c_uint8 * 4),
+                     ('d_name', ctypes.c_char * 256),
+                 )
++            elif 'darwin' in sys.platform and 'arm64' in platform.machine():
++                _fields_ = (
++                    ('d_ino', ctypes.c_uint64),
++                    ('d_off', ctypes.c_uint64),
++                    ('d_reclen', ctypes.c_uint16),
++                    ('d_namlen', ctypes.c_uint16),
++                    ('d_type', ctypes.c_uint8),
++                    ('d_name', ctypes.c_char * 1024),
++                )
+             else:
+                 _fields_ = (
+                     ('d_ino', ctypes.c_uint32),  # must be uint32, not ulong

--- a/pkgs/development/python2-modules/scandir/default.nix
+++ b/pkgs/development/python2-modules/scandir/default.nix
@@ -6,8 +6,12 @@ buildPythonPackage rec {
 
   src = fetchPypi {
     inherit pname version;
-    sha256 ="1bkqwmf056pkchf05ywbnf659wqlp6lljcdb0y88wr9f0vv32ijd";
+    sha256 = "1bkqwmf056pkchf05ywbnf659wqlp6lljcdb0y88wr9f0vv32ijd";
   };
+
+  patches = [
+    ./add-aarch64-darwin-dirent.patch
+  ];
 
   checkPhase = "${python.interpreter} test/run_tests.py";
 

